includes/class-amp-theme-support.php

```diff
@@ -458,6 +471,18 @@ public static function ensure_proper_amp_location( $exit = true ) {
 			if ( $has_query_var || $has_url_param ) {
 				return self::redirect_non_amp_url( current_user_can( 'manage_options' ) ? 302 : 301, $exit );
 			}
+		} elseif ( self::READER_MODE_SLUG === self::get_support_mode() && is_singular() ) {
+			// Prevent infinite URL space under /amp/ endpoint.
+			global $wp;
+			$path_args = [];
+			wp_parse_str( $wp->matched_query, $path_args );
+			if ( isset( $path_args[ amp_get_slug() ] ) && '' !== $path_args[ amp_get_slug() ] ) {
+				wp_safe_redirect( amp_get_permalink( get_queried_object_id() ), 301 );
+				if ( $exit ) {
+					exit;
+				}
+				return true;
+			}
 		} else {
```

Comment:
There's a bogus if below here that can be included in the else.
```suggestion
		} elseif ( $has_query_var && ! $has_url_param ) {
```

Reply:
Fixed in 93cede44d26e0bdadbc3fa99db0ed6cedbfad36d